packages/safe-ds-lang/src/language/custom-editor/ast-parser/utils.ts

```diff
@@ -0,0 +1,129 @@
+        Utils.documentUri = documentUri ? documentUri : Utils.documentUri;
+        Utils.safeDsServices = safeDsServices
+            ? safeDsServices
+            : Utils.safeDsServices;
+    }
+
+    private static constructErrorMessage(message: string, origin?: AstNode) {
+        const uri = origin?.$cstNode?.root.astNode.$document?.uri.fsPath ?? "";
+        const position = origin?.$cstNode
+            ? `:${origin.$cstNode.range.start.line + 1}:${origin.$cstNode.range.start.character + 1}`
+            : "";
+
+        return `${uri}${position} - ${message}`;
+    }
+
+    public static log(message: string) {
+        Utils.logger.debug("AstParser", message);
+    }
+
+    public static logError(message: string, origin?: AstNode) {
+        Utils.logger.error(
+            "AstParser",
+            origin ? Utils.constructErrorMessage(message, origin) : message,
+        );
+    }
+
+    public static pushError(message: string, origin?: AstNode) {
+        const error = new CustomError(
+            "block",
+            Utils.constructErrorMessage(message, origin),
+        );
+        Utils.errorList.push(error);
+        Utils.logError(error.message);
+        return error;
+    }
+
+    public static pushLexerErrors(error: ILexingError) {
+        const uri = Utils.documentUri.toString();
+        const position =
+            error.line && error.column
+                ? `:${error.line + 1}:${error.column + 1}`
+                : "";
+
+        const message = `${uri}${position} - Lexer Error: ${error.message}`;
+
+        const fullError = `${uri}${position} - ${message}`;
+        Utils.errorList.push(new CustomError("block", fullError));
+        Utils.logError(message);
+    }
+
+    public static pushParserErrors(error: IRecognitionException) {
+        const uri = Utils.documentUri.toString();
+        const position =
+            error.token.startLine && error.token.startColumn
+                ? `:${error.token.startLine + 1}:${error.token.startColumn + 1}`
+                : "";
+
+        const message = `${uri}${position} - Parser Error: ${error.message}`;
+
+        const fullError = `${uri}${position} - ${message}`;
+        Utils.errorList.push(new CustomError("block", fullError));
+        Utils.logError(message);
+    }
+
+    public static getNewId() {
+        return Utils.lastId++;
+    }
+
+    public static collectAst(): Ast {
+        const ast = {
+            placeholderList: Utils.placeholderList,
+            callList: Utils.callList,
+            genericExpressionList: Utils.genericExpressionList,
+            edgeList: Utils.edgeList,
+        };
+        return ast;
+    }
+}
+
+export const zip = <A, B>(arrayA: A[], arrayB: B[]): [A, B][] => {
+    const minLength = Math.min(arrayA.length, arrayB.length);
+    const result: [A, B][] = [];
+
+    for (let i = 0; i < minLength; i++) {
+        result.push([arrayA[i]!, arrayB[i]!]);
+    }
+
+    return result;
+};
```

Comment:
Would be a useful addition to `packages\safe-ds-lang\src\helpers\collections.ts`.